src/main/java/com/guysfromusa/carsgame/control/MessageDispatcher.java

```diff
@@ -41,38 +46,42 @@ public void init() {
 
     @Override
     public void run() {
-        while (true) {
-            log.info("Consume messages from queue");
-
-            //TODO temporary solution, refactor me
-            //TODO filter messages by type and game and forward to engine
-            Message firstMessage = null;
+        while(true) {
             try {
-                firstMessage = gameQueue.take();
-                log.info("Consumed first message");
-            } catch (InterruptedException e) {
-                e.printStackTrace();
+                queuesNotEmptyBarrier.await();
+            } catch (InterruptedException | BrokenBarrierException e) {
+                log.error("Barrier terminated: ", e);
             }
-            List<Message> consumedMessages = new ArrayList<>(singletonList(firstMessage));
-            log.info("DrainTo for messages");
-            gameQueue.drainTo(consumedMessages);
 
+            Optional<GameState> gameToPlayRoundOptional = gameStateTracker.getGameStates().stream()
+                    .filter(state -> !state.isRoundInProgress())
+                    .filter(state -> !state.getMovementsQueue().isEmpty())
+                    .findFirst();
 
-            Map<MessageType, List<Message>> byType = consumedMessages.stream()
-                    .collect(groupingBy(Message::getMessageType));
+            gameToPlayRoundOptional.ifPresent(gameState -> {
+                gameState.setRoundInProgress(true);
 
-            byType.forEach(this::handle);
+                //FIXME filter messages by type and group different cars
+                List<Message> consumedMessages = new ArrayList<>();
+                gameState.getMovementsQueue().drainTo(consumedMessages);
+                handle(MOVE, consumedMessages, gameState.getGameName());
+            });
+
+            if (!gameToPlayRoundOptional.isPresent()) {
+                queuesNotEmptyBarrier.reset();
+            }
         }
     }
 
-    private void handle(MessageType messageType, List<Message> messages) {
+    private void handle(MessageType messageType, List<Message> messages, String gameName) {
         //TODO Strategy someday
         switch (messageType) {
             case MOVE:
-                gameEngine.handleMoves(messages);
+                gameEngine.handleMoves(messages, gameName);
                 break;
             default:
                 throw new IllegalStateException("Undefined message type");
         }
+        gameStateTracker.getGameState(gameName).setRoundInProgress(false);
```

Comment:
this is a bug, since handleMoves is Async we should call this line as a last step of handleMoves and additionally trigger there an event/release barrier to recheck the queues since maybe there will be message not filtered out by 

!state.isRoundInProgress()